Add Change Background Feature @Sketch Page #634

nijin26 opened the pull request:
## Issue #634 is DONE

- [x] Change background colour using Color picker.
- [x] Save/Download that canvas as an Image with that specific colour as the background.
- [x] It's possible to change colour before sketching.
- [x] Changing canvas colour after sketching may cause overlay of colour above shape, strokes or text.    

### Sketch Page
![bgcolor change](https://user-images.githubusercontent.com/63832147/111303629-bd00a400-867a-11eb-927a-0479df8b3410.png)

### Image of the above Canvas
![drawing (35)](https://user-images.githubusercontent.com/63832147/111303687-cee24700-867a-11eb-9fa6-19ea843a8810.png)

iamdarshshah commented on src/pages/Sketch/components/Canvas.js:
```diff
@@ -139,31 +140,31 @@ function Canvas() {
+        } else if (type === 'line') {
             lineMove(point);
-        } else if(type === 'square') {
+        } else if (type === 'square') {
             squareMove(point);
-        } else if(type === 'circle') {
+        } else if (type === 'circle') {
             circleMove(point);
-        } else if(type === 'triangle') {
+        } else if (type === 'triangle') {
             triangleMove(point);
-        } else if(type === 'arrow') {
+        } else if (type === 'arrow') {
             arrow(point);
-        } else if(type === 'diamond') {
+        } else if (type === 'diamond') {
             diamondMove(point);
         }
```
Use Switch instead.

iamdarshshah replied:
@smaranjitghose It's better to use Switch for this scenario, any thoughts?

iamdarshshah commented on src/pages/Sketch/components/Canvas.js:
```diff
@@ -100,29 +101,29 @@ function Canvas() {
+        if (type === 'pen') {
             logicDown(point);
-        } else if(type === 'line' || type === 'square' || type === 'circle' || type === 'triangle' || type === 'arrow' || type === 'diamond') {
+        } else if (type === 'line' || type === 'square' || type === 'circle' || type === 'triangle' || type === 'arrow' || type === 'diamond') {
```
```suggestion
        } else if (['line', 'square', 'circle', 'triangle', 'arrow', 'diamond'].includes(type)) {
```
DO this instead.

iamdarshshah replied:
@smaranjitghose Don't you think this is the better way?

nijin26 replied:
Ok Sir

smaranjitghose commented on src/pages/Sketch/components/Toolbox/Toolbox.js:
```diff
@@ -41,6 +43,15 @@ function Toolbox({
 
+            <Feature title="Background">
+                <div className={styles.featureBg}>
+                <small>"Set it Before Sketching"</small>
```
This line is not at all required. Please remove it

smaranjitghose commented on src/pages/Sketch/components/Toolbox/Toolbox.js:
```diff
@@ -41,6 +43,15 @@ function Toolbox({
                 />
             </Feature>
 
+            <Feature title="Background">
```
Make this ``Canvas Color``

nijin26 replied:
Ok

smaranjitghose commented:
Along with the color picker, try to incorporate getting colors using hexcode rather than RGB values

iamdarshshah commented:
@smaranjitghose Don't you think it's good if we add the suggested changes and then mark it as PR-approved?

smaranjitghose commented:
> @smaranjitghose Don't you think it's good if we add the suggested changes and then mark it as PR-approved?

Ya.. This github action is messing up😅

iamdarshshah commented:
@smaranjitghose Oh Okay, No worries! It seems we do have to make certain changes in Github actions workflow 👀

nijin26 commented:
## Description:

* Hex input added along with Color picker(for both Canvas Color & Color)
* New feature's name changed from "Background" to "Canvas Color"

![Screenshot from 2021-03-16 23-25-07](https://user-images.githubusercontent.com/63832147/111356952-dd4a5600-86ae-11eb-83f8-f596a5ada8de.png)

nijin26 commented:
Updated Code as per @iamdarshshah 's suggestion.

* Used Switch instead of a bunch of if....else in line 155
* Added every strings to an array and used includes method instead of OR, in line 118

iamdarshshah commented on src/pages/Sketch/components/Canvas.js:
```diff
@@ -40,58 +41,58 @@ function Canvas() {
 
         return () => {
-            window.removeEventListener('resize', () => {});
+            window.removeEventListener('resize', () => { });
```
```suggestion
            window.removeEventListener('resize', handleResizeListener);
```
This is not the correct way to remove the event listeners. You'll have to create `handleResizeListener` (or name any) function outside of useEffect scope and this will be a callback function that you've to pass while adding an event listener (i.e, in line number 51) as well as while removing the event listener.

nijin26 replied:
Sir, these are already existed code , done by other contributers.

Anyway thanks for your suggestions, I will make the change.

iamdarshshah replied:
@nijin26 I understand but better to make it look more readable and it's always a good practice to update the previously created code and refactor it in a way that will make it more efficient. 

P.S Just a suggestion: You have to make certain changes in the codebase (at least the files you're working on) irrespective of whether you've worked on it or others might have worked on that block of code, just to handle certain things in right way. 🙂

nijin26 replied:
Yes Sir,
I will do that .😇

iamdarshshah commented on src/pages/Sketch/components/Canvas.js:
```diff
@@ -40,58 +41,58 @@ function Canvas() {
 
     useEffect(() => {
         window.addEventListener('resize', () => {
-            setCanvasWidth(window.innerWidth-50);
-            setCanvasHeight(window.innerHeight-100);
+            setCanvasWidth(window.innerWidth - 50);
+            setCanvasHeight(window.innerHeight - 100);
         })
```
```
const handleResizeListener = () => {
            setCanvasWidth(window.innerWidth - 50);
            setCanvasHeight(window.innerHeight - 100);
        }
 useEffect(() => {
        window.addEventListener('resize', handleResizeListener)
```
You can do it this way.

iamdarshshah reviewed (approved):
LGTM! 🎉

nijin26 commented:
@anushbhatia @iamdarshshah 

Some PRs got nerged above this one and canvas colour option is currently missing from live site. 😑

iamdarshshah commented:
@nijin26 Can you please raise an issue for the same? Also, it would be great if you can add some analysis about what might've caused this issue. 

Thanks.
cc: @anushbhatia @smaranjitghose
